ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FantomasHost.fs

```diff
@@ -20,20 +20,29 @@ module internal Reflection =
 
 
 [<SolutionComponent>]
-type FantomasHost(solution: ISolution, fantomasFactory: FantomasProcessFactory) =
+type FantomasHost(solution: ISolution, fantomasFactory: FantomasProcessFactory, fantomasDetector: FantomasDetector) =
+    let solutionLifetime = solution.GetLifetime()
     let mutable connection: FantomasConnection = null
     let mutable formatConfigFields: string[] = [||]
+    let mutable formatterHostLifetime: LifetimeDefinition = null
 
     let toEditorConfigName name = $"{fSharpEditorConfigPrefix}{StringUtil.MakeUnderscoreCaseName(name)}"
 
     let isConnectionAlive () =
         isNotNull connection && connection.IsActive
 
+    let terminateConnection () =
+        if isConnectionAlive () then formatterHostLifetime.Terminate()
+
+    // Formatting is called on UI thread, so we don't need to synchronize it
```

Comment:
I think it worth it to either synchronize it or to assert we only access Fantomas on the main thread. Asserting is definitely a better way to make it safe than a comment. Otherwise it'd be a hard to spot issue if something changes in future (and I do hope we'll have less things working on the main thread like this).

Reply:
Generally speaking, we have all the necessary synchronization in `FantomasDetector`, through which we notify about a Fantomas version change and run Fantomas. If we wrap the entire body of the `connect` function in fantomasDetector.Run(...), then have all the necessary synchronization regardless of the thread calling formating.